hack/generator/pkg/codegen/testdata/RecursiveTypes/Unrolls_recursive_type_multiple_loops.golden

```diff
@@ -0,0 +1,20 @@
+// Code generated by azure-service-operator-codegen. DO NOT EDIT.
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+package v1alpha1api20200101
+
+//Generated from: https://test.test/schemas/2020-01-01/test.json
+type Test struct {
+	Error *Error `json:"error,omitempty"`
+}
+
+//Generated from: https://test.test/schemas/2020-01-01/test.json#/definitions/Error
+type Error struct {
+	Error1 *Error_Unrolled `json:"error1,omitempty"`
+	Error2 *Error_Unrolled `json:"error2,omitempty"`
+	Name   *string         `json:"name,omitempty"`
+}
```

Comment:
I'm not sure I follow - they're already JSON annotated and should already deserialize from what ARM is returning to us. All I did was limit the recursive depth.

So the following will deserialize fine:
```json
{
    "name": "Outer error",
    "error1": {
        "name": "Inner error 1"
    },
    "error2": {
        "name": "Inner error 2"
    }
}
```

Whereas this (which is allowed as per the Swagger) will not anymore:
```json
{
    "name": "Outer error",
    "error1": {
        "name": "Inner error 1",
        "error1": {
            "name": "inner inner error 1"
        }
    },
    "error2": {
        "name": "Inner error 2",
        "error2": {
            "name": "inner inner error 2"
        }
    }
}
```

You can see where this comes up in the [ARM common API definitions](https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/common-api-details.md#error-response-content). The actual ARM error types usually look like [this](https://github.com/Azure/azure-rest-api-specs/blob/master/specification/common-types/resource-management/v2/types.json#L285) - note that a lot of specs (Batch, etc) don't use the same type but instead redefine a type with a similar structure (which is why we can't just name a single type to unroll and be done here).

So the answer to "will it deserialize" is yes. We represent a subset of what's actually allowed by the API so it will always deserialize _something_. The harder question is, will that something be *complete*, and that boils down to if services actually in practice nest details inside their error details (so 3+ levels instead of 2). My observation is that they don't and that aligns with ARMs example, but obviously the type they've defined is broader than that so there's no guarantee.

Reply:
Thanks for clarifying.

Reply:
Discussed this and decided to only do this for things with `Error` in the name, and also log an error or warning if we run into one that doesn't match that description.